algorithms/arrays/n_sum.py

```diff
@@ -0,0 +1,129 @@
+"""
+Given an array of n integers, are there elements a, b, .. , n in nums
+such that a + b + .. + n = target?
+
+Find all unique triplets in the array which gives the sum of target.
+
+Example:
+    basic:
+        Given:
+            n = 4, nums = [1, 0, -1, 0, -2, 2], target = 0,
+        return [[-2, -1, 1, 2], [-2, 0, 0, 2], [-1, 0, 0, 1]]
+
+    advanced:
+        Given:
+            n = 2
+            nums = [[-3, 0], [-2, 1], [2, 2], [3, 3], [8, 4], [-9, 5]]
+            taget = -5
+            def sum(a, b):
+                return [a[0] + b[1], a[1] + b[0]]
+            def compare(num, taget):
+                if num[0] < taget:
+                    return -1
+                elif if num[0] > taget:
+                    return 1
+                else:
+                    return 0
+        return [[-9, 5], [8, 4]]
+        because -9 + 4 = -5
+"""
+
+
+def n_sum(n, nums, target, **kv):
+    """
+    n: int
+    nums: list[object]
+    target: object
+    sum_closure: function, optional
+                 Given two elements of nums, return sum of both.
+    compare_closure: function, optional
+                     Given one object of nums and target, return one of -1, 1, or 0.
+    same_closure: function, optional
+                  Given two object of nums, return bool.
+    return: list[list[object]]
+
+    Note:
+    1. type of sum_closure's return should be same as type of compare_closure's first param
+    """
+
+    def sum_closure_default(a, b):
+        return a + b
+
+    def compare_closure_default(num, taget):
+        if num < taget:
+            return -1
+        elif num > taget:
+            return 1
+        else:
+            return 0
+
+    def same_closure_default(a, b):
+        return a == b
+
+    def n_sum(n, nums, target):
```

Comment:
Wow, I am surprised we can define inner function with the same name as outer. I didn't know it was possible in Python, but it works! Thanks.